### PR TITLE
Adding random number generator to base dataloader for random access mode

### PR DESCRIPTION
**Pull request by evsmithx**
Also added serialisation for the rng.

### CODE REVIEW

**Review by baykaner — DISMISSED** [review]

File: libs/ml/tests/unit/ops/activations/dropout.cpp [hunk 1]
```diff
@@ -54,137 +70,114 @@ TYPED_TEST(DropoutTest, forward_test)
 
   op.Forward(vec_data, prediction);
 
-  // test correct values
-  EXPECT_TRUE(prediction.AllClose(gt, DataType{1e-5f}, DataType{1e-5f}));
+  // test correct fraction and sum
+  EXPECT_NEAR(zero_fraction(prediction), static_cast<double>(prob), 0.05);
```
Comment (baykaner) [inline libs/ml/tests/unit/ops/activations/dropout.cpp]: 0.05 is presumably a slightly arbitrary value which is contingent on the size of the tensor? we could point this out with a comment

**Review by baykaner — DISMISSED** [review]

File: libs/ml/tests/unit/clustering/tsne_tests.cpp [hunk 2]
```diff
@@ -104,48 +104,24 @@ TYPED_TEST(TsneTests, tsne_test_2d)
...
               50 * static_cast<double>(math::function_tolerance<DataType>()));
-  EXPECT_NEAR(double(output_matrix.At(0, 99)), -0.87261968106031417846679687500000,
+  EXPECT_NEAR(double(output_matrix.At(0, 99)), 2.7302324351584537,
               50 * static_cast<double>(math::function_tolerance<DataType>()));
-  EXPECT_NEAR(double(output_matrix.At(1, 99)), 3.04633120773360133171081542968750,
+  EXPECT_NEAR(double(output_matrix.At(1, 99)), 0.48101261687371411,
               50 * static_cast<double>(math::function_tolerance<DataType>()));
 }
 
-TYPED_TEST(TsneTests, tsne_test_2d_cross_type_consistency_test)
```
Comment (MissingNO57) [inline libs/ml/tests/unit/clustering/tsne_tests.cpp:125]: Why is this test removed?

**Review by MissingNO57 — APPROVED** [review]
Very nice!

File: libs/ml/tests/unit/clustering/tsne_tests.cpp (hunk 2 above)
Comment (evsmithx) [reply libs/ml/tests/unit/clustering/tsne_tests.cpp:125]: As discussed, it's a duplicate of the test above!

**Pull request merged**